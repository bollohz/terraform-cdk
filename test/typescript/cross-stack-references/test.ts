import * as fs from "fs-extra";
import * as path from "path";
import { TestDriver } from "../../test-helper";

describe("cross stack references", () => {
  let driver: TestDriver;

  beforeAll(async () => {
    driver = new TestDriver(__dirname);
    await driver.setupTypescriptProject();
    await driver.synth();
  });

  test("synth generates JSON", () => {
    expect(driver.manifest()).toMatchSnapshot();
  });

  describe("deployed", () => {
    beforeAll(async () => {
      await driver.deploy("origin");
      await driver.deploy("passthrough");
      await driver.deploy("sink");
    });

    it("references primitive values", () => {
      expect(driver.readLocalFile("originNum")).toBe(
        driver.readLocalFile("passthroughNum")
      );
      expect(driver.readLocalFile("originStr")).toBe(
        driver.readLocalFile("passthroughStr")
      );
    });

    it("references can be passed through stacks", () => {
      expect(driver.readLocalFile("originNum")).toBe(
        driver.readLocalFile("sinkNum")
      );
      expect(driver.readLocalFile("originStr")).toBe(
        driver.readLocalFile("sinkStr")
      );
    });

    it("references complex values", () => {
      expect(driver.readLocalFile("originList")).toBe(
        driver.readLocalFile("passthroughList")
      );
    });

    describe("terraform functions", () => {
      beforeAll(async () => {
        await driver.deploy("fns");
        await driver.deploy("functionOutput");
      });

      it("can use reference in terraform function", () => {
        // num: Fn.sum([src.num, src.num]),
        const originNum = parseInt(driver.readLocalFile("originNum"), 10);
        const result = parseInt(driver.readLocalFile("fnsNum"), 10);

        expect(originNum * 2).toBe(result);
      });

      it("can use references from two stacks", () => {
        // str: Fn.join(",", [src.str, passthrough.str]),
        const originStr = driver.readLocalFile("originStr");
        const passthroughStr = driver.readLocalFile("passthroughStr");

        const result = driver.readLocalFile("fnsStr");
        expect(`${originStr},${passthroughStr}`).toBe(result);
      });

      it("references terraform function output", () => {
        const originNum = parseInt(driver.readLocalFile("originNum"), 10);
        const result = parseInt(driver.readLocalFile("functionOutputNum"), 10);

        expect(originNum * 2).toBe(result);
      });
    });

    // Check for Deadly embrace scenario: https://github.com/aws/aws-cdk/pull/12778
    it("can remove references to deployed stacks", async () => {
      driver.setEnv("SWITCH_STACK", "on");
      console.log(driver.workingDirectory);
      await driver.deploy("secondOrigin");
      await driver.deploy("switchedStack");
      expect(driver.manifest()).toMatchSnapshot("with dependency");

      driver.setEnv("SWITCH_STACK", undefined);
      await driver.deploy("secondOrigin");
      await driver.deploy("switchedStack");
      expect(driver.manifest()).toMatchSnapshot("without dependency");
    });

    it("can pin function outputs to a stack by using terraform locals", async () => {
      await driver.deploy("pinnedFns");
      await driver.deploy("functionOutputPinned");

      const pinnedStr = driver.readLocalFile("pinnedFnsStr");
      const outputPinnedStr = driver.readLocalFile("functionOutputPinnedStr");
      expect(pinnedStr).toEqual(outputPinnedStr);
    });
  });
});
