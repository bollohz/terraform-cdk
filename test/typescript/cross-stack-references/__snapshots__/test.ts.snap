// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cross stack references deployed can remove references to deployed stacks: with dependency 1`] = `
"{
  \\"version\\": \\"stubbed\\",
  \\"stacks\\": {
    \\"origin\\": {
      \\"name\\": \\"origin\\",
      \\"constructPath\\": \\"origin\\",
      \\"workingDirectory\\": \\"stacks/origin\\",
      \\"synthesizedStackPath\\": \\"stacks/origin/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": []
    },
    \\"passthrough\\": {
      \\"name\\": \\"passthrough\\",
      \\"constructPath\\": \\"passthrough\\",
      \\"workingDirectory\\": \\"stacks/passthrough\\",
      \\"synthesizedStackPath\\": \\"stacks/passthrough/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": [
        \\"origin\\"
      ]
    },
    \\"sink\\": {
      \\"name\\": \\"sink\\",
      \\"constructPath\\": \\"sink\\",
      \\"workingDirectory\\": \\"stacks/sink\\",
      \\"synthesizedStackPath\\": \\"stacks/sink/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": [
        \\"origin\\"
      ]
    },
    \\"fns\\": {
      \\"name\\": \\"fns\\",
      \\"constructPath\\": \\"fns\\",
      \\"workingDirectory\\": \\"stacks/fns\\",
      \\"synthesizedStackPath\\": \\"stacks/fns/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": [
        \\"origin\\"
      ]
    },
    \\"functionOutput\\": {
      \\"name\\": \\"functionOutput\\",
      \\"constructPath\\": \\"functionOutput\\",
      \\"workingDirectory\\": \\"stacks/functionOutput\\",
      \\"synthesizedStackPath\\": \\"stacks/functionOutput/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": [
        \\"origin\\"
      ]
    },
    \\"pinnedFns\\": {
      \\"name\\": \\"pinnedFns\\",
      \\"constructPath\\": \\"pinnedFns\\",
      \\"workingDirectory\\": \\"stacks/pinnedFns\\",
      \\"synthesizedStackPath\\": \\"stacks/pinnedFns/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": [
        \\"origin\\"
      ]
    },
    \\"functionOutputPinned\\": {
      \\"name\\": \\"functionOutputPinned\\",
      \\"constructPath\\": \\"functionOutputPinned\\",
      \\"workingDirectory\\": \\"stacks/functionOutputPinned\\",
      \\"synthesizedStackPath\\": \\"stacks/functionOutputPinned/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": [
        \\"pinnedFns\\"
      ]
    },
    \\"secondOrigin\\": {
      \\"name\\": \\"secondOrigin\\",
      \\"constructPath\\": \\"secondOrigin\\",
      \\"workingDirectory\\": \\"stacks/secondOrigin\\",
      \\"synthesizedStackPath\\": \\"stacks/secondOrigin/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": []
    },
    \\"switchedStack\\": {
      \\"name\\": \\"switchedStack\\",
      \\"constructPath\\": \\"switchedStack\\",
      \\"workingDirectory\\": \\"stacks/switchedStack\\",
      \\"synthesizedStackPath\\": \\"stacks/switchedStack/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": [
        \\"secondOrigin\\"
      ]
    }
  }
}"
`;

exports[`cross stack references deployed can remove references to deployed stacks: without dependency 1`] = `
"{
  \\"version\\": \\"stubbed\\",
  \\"stacks\\": {
    \\"origin\\": {
      \\"name\\": \\"origin\\",
      \\"constructPath\\": \\"origin\\",
      \\"workingDirectory\\": \\"stacks/origin\\",
      \\"synthesizedStackPath\\": \\"stacks/origin/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": []
    },
    \\"passthrough\\": {
      \\"name\\": \\"passthrough\\",
      \\"constructPath\\": \\"passthrough\\",
      \\"workingDirectory\\": \\"stacks/passthrough\\",
      \\"synthesizedStackPath\\": \\"stacks/passthrough/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": [
        \\"origin\\"
      ]
    },
    \\"sink\\": {
      \\"name\\": \\"sink\\",
      \\"constructPath\\": \\"sink\\",
      \\"workingDirectory\\": \\"stacks/sink\\",
      \\"synthesizedStackPath\\": \\"stacks/sink/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": [
        \\"origin\\"
      ]
    },
    \\"fns\\": {
      \\"name\\": \\"fns\\",
      \\"constructPath\\": \\"fns\\",
      \\"workingDirectory\\": \\"stacks/fns\\",
      \\"synthesizedStackPath\\": \\"stacks/fns/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": [
        \\"origin\\"
      ]
    },
    \\"functionOutput\\": {
      \\"name\\": \\"functionOutput\\",
      \\"constructPath\\": \\"functionOutput\\",
      \\"workingDirectory\\": \\"stacks/functionOutput\\",
      \\"synthesizedStackPath\\": \\"stacks/functionOutput/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": [
        \\"origin\\"
      ]
    },
    \\"pinnedFns\\": {
      \\"name\\": \\"pinnedFns\\",
      \\"constructPath\\": \\"pinnedFns\\",
      \\"workingDirectory\\": \\"stacks/pinnedFns\\",
      \\"synthesizedStackPath\\": \\"stacks/pinnedFns/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": [
        \\"origin\\"
      ]
    },
    \\"functionOutputPinned\\": {
      \\"name\\": \\"functionOutputPinned\\",
      \\"constructPath\\": \\"functionOutputPinned\\",
      \\"workingDirectory\\": \\"stacks/functionOutputPinned\\",
      \\"synthesizedStackPath\\": \\"stacks/functionOutputPinned/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": [
        \\"pinnedFns\\"
      ]
    },
    \\"secondOrigin\\": {
      \\"name\\": \\"secondOrigin\\",
      \\"constructPath\\": \\"secondOrigin\\",
      \\"workingDirectory\\": \\"stacks/secondOrigin\\",
      \\"synthesizedStackPath\\": \\"stacks/secondOrigin/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": []
    },
    \\"switchedStack\\": {
      \\"name\\": \\"switchedStack\\",
      \\"constructPath\\": \\"switchedStack\\",
      \\"workingDirectory\\": \\"stacks/switchedStack\\",
      \\"synthesizedStackPath\\": \\"stacks/switchedStack/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": []
    }
  }
}"
`;

exports[`cross stack references synth generates JSON 1`] = `
"{
  \\"version\\": \\"stubbed\\",
  \\"stacks\\": {
    \\"origin\\": {
      \\"name\\": \\"origin\\",
      \\"constructPath\\": \\"origin\\",
      \\"workingDirectory\\": \\"stacks/origin\\",
      \\"synthesizedStackPath\\": \\"stacks/origin/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": []
    },
    \\"passthrough\\": {
      \\"name\\": \\"passthrough\\",
      \\"constructPath\\": \\"passthrough\\",
      \\"workingDirectory\\": \\"stacks/passthrough\\",
      \\"synthesizedStackPath\\": \\"stacks/passthrough/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": [
        \\"origin\\"
      ]
    },
    \\"sink\\": {
      \\"name\\": \\"sink\\",
      \\"constructPath\\": \\"sink\\",
      \\"workingDirectory\\": \\"stacks/sink\\",
      \\"synthesizedStackPath\\": \\"stacks/sink/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": [
        \\"origin\\"
      ]
    },
    \\"fns\\": {
      \\"name\\": \\"fns\\",
      \\"constructPath\\": \\"fns\\",
      \\"workingDirectory\\": \\"stacks/fns\\",
      \\"synthesizedStackPath\\": \\"stacks/fns/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": [
        \\"origin\\"
      ]
    },
    \\"functionOutput\\": {
      \\"name\\": \\"functionOutput\\",
      \\"constructPath\\": \\"functionOutput\\",
      \\"workingDirectory\\": \\"stacks/functionOutput\\",
      \\"synthesizedStackPath\\": \\"stacks/functionOutput/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": [
        \\"origin\\"
      ]
    },
    \\"pinnedFns\\": {
      \\"name\\": \\"pinnedFns\\",
      \\"constructPath\\": \\"pinnedFns\\",
      \\"workingDirectory\\": \\"stacks/pinnedFns\\",
      \\"synthesizedStackPath\\": \\"stacks/pinnedFns/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": [
        \\"origin\\"
      ]
    },
    \\"functionOutputPinned\\": {
      \\"name\\": \\"functionOutputPinned\\",
      \\"constructPath\\": \\"functionOutputPinned\\",
      \\"workingDirectory\\": \\"stacks/functionOutputPinned\\",
      \\"synthesizedStackPath\\": \\"stacks/functionOutputPinned/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": [
        \\"pinnedFns\\"
      ]
    },
    \\"secondOrigin\\": {
      \\"name\\": \\"secondOrigin\\",
      \\"constructPath\\": \\"secondOrigin\\",
      \\"workingDirectory\\": \\"stacks/secondOrigin\\",
      \\"synthesizedStackPath\\": \\"stacks/secondOrigin/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": []
    },
    \\"switchedStack\\": {
      \\"name\\": \\"switchedStack\\",
      \\"constructPath\\": \\"switchedStack\\",
      \\"workingDirectory\\": \\"stacks/switchedStack\\",
      \\"synthesizedStackPath\\": \\"stacks/switchedStack/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": []
    }
  }
}"
`;
